src/Symfony/Component/DependencyInjection/Attribute/AutowireIterator.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Attribute;
+
+use Symfony\Component\DependencyInjection\Argument\IteratorArgument;
+use Symfony\Component\DependencyInjection\Argument\TaggedIteratorArgument;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\TypedReference;
+use Symfony\Contracts\Service\Attribute\SubscribedService;
+use Symfony\Contracts\Service\ServiceSubscriberInterface;
+
+#[\Attribute(\Attribute::TARGET_PARAMETER)]
+class AutowireIterator extends Autowire
+{
+    /**
+     * @see ServiceSubscriberInterface::getSubscribedServices()
+     *
+     * @param array<string|SubscribedService>|string $services An explicit list of services or a tag name
+     */
+    public function __construct(
+        array|string $services,
+        string $indexAttribute = null,
+        string $defaultIndexMethod = null,
+        string $defaultPriorityMethod = null,
+        string|array $exclude = [],
+        bool $excludeSelf = true,
+    ) {
+        if (\is_string($services)) {
+            parent::__construct(new TaggedIteratorArgument($services, $indexAttribute, $defaultIndexMethod, false, $defaultPriorityMethod, (array) $exclude, $excludeSelf));
+
+            return;
+        }
+
+        $references = [];
+
+        foreach ($services as $key => $type) {
+            $attributes = [];
+
+            if ($type instanceof SubscribedService) {
+                $key = $type->key ?? $key;
+                $attributes = $type->attributes;
+                $type = ($type->nullable ? '?' : '').($type->type ?? throw new InvalidArgumentException(sprintf('When "%s" is used, a type must be set.', SubscribedService::class)));
+            }
+
+            if (!\is_string($type) || !preg_match('/(?(DEFINE)(?<cn>[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+))(?(DEFINE)(?<fqcn>(?&cn)(?:\\\\(?&cn))*+))^\??(?&fqcn)(?:(?:\|(?&fqcn))*+|(?:&(?&fqcn))*+)$/', $type)) {
+                throw new InvalidArgumentException(sprintf('"%s" is not a PHP type for key "%s".', \is_string($type) ? $type : get_debug_type($type), $key));
+            }
+            if ($optionalBehavior = '?' === $type[0]) {
```

Comment:
I know you like to make assignments in `if`, but in this case it's much more readable if extracted (or use `else`). The `TypedReference` creation can also be simplified.
```suggestion
            $optionalBehavior = ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE;
            if ('?' === $type[0]) {
```

Reply:
that would break the `?:  below though. Maybe rename it to `$invalidBehavior` and use a good old `else` instead (I agree this one is hard to read)

Reply:
Would mean the same :) Also, it's borrowed from RegisterServiceSubscribersPass

Reply:
I understand it was done this way 6 years ago in [`RegisterServiceSubscribersPass`](https://github.com/symfony/symfony/blame/d27190a4cf8560b77d8aaab65104e51a2dd5c06c/src/Symfony/Component/DependencyInjection/Compiler/RegisterServiceSubscribersPass.php#L90-L93). But let me insist, I don't see why the implementation "Wouldn't mean the same".

```php
$optionalBehavior = ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE;
if ('?' === $type[0]) {
    $type = substr($type, 1);
    $optionalBehavior = ContainerInterface::IGNORE_ON_INVALID_REFERENCE;
}
// ...
$references[$key] = new TypedReference($type, $type, $optionalBehavior, $name, $attributes);
```

Reply:
alright, updated!